WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountViewModel.swift

```diff
@@ -4,53 +4,84 @@ import UIKit
 import SwiftUI
 import Yosemite
 
-typealias CustomAmountEntered = (_ amount: String, _ name: String, _ feeID: Int64?, _ isTaxable: Bool) -> Void
+/// Provides a adapter depending on the input type
+///
+final class AddCustomAmountInputTypeViewModelAdapterProvider {
+    func provideAddCustomAmountInputTypeViewModelAdapter(with type: AddCustomAmountViewModel.InputType) -> AddCustomAmountInputTypeViewModelAdapter {
+        switch type {
+        case .fixedAmount:
+            return FixedAmountAddCustomAmountInputTypeViewModelAdapter()
+        case .orderTotalPercentage(let baseAmount):
+            return PercentageAddCustomAmountInputTypeViewModelAdapter(baseAmount: baseAmount)
+        }
+    }
+}
 
-final class AddCustomAmountViewModel: ObservableObject {
-    let formattableAmountTextFieldViewModel: FormattableAmountTextFieldViewModel
-    private var baseAmountForPercentage: Decimal
-    private let onCustomAmountEntered: CustomAmountEntered
-    private let analytics: Analytics
-    private let currencyFormatter: CurrencyFormatter
+/// Classes implementing this protocol will act as a adapter between `AddCustomAmountViewModel` and the input type view models.
+/// This way, the former can be agnostic of what type of view model we have before the scenes,
+/// and doesn't have to type check to access their concrete implementation.
+/// Having a adapter keeps the classes loosely coupled, as the type view model doesn't have to adapt for `AddCustomAmountViewModel`,
+/// the adapter implements this adaptation.
+///
+protocol AddCustomAmountInputTypeViewModelAdapter {
+    var amount: String? { get }
+    var amountPublisher: Published<String>.Publisher? { get }
+    var viewModel: AddCustomAmountViewModel? { get set }
+
+    func preset(with fee: OrderFeeLine)
+}
 
-    var baseAmountForPercentageString: String {
-        currencyFormatter.formatAmount(baseAmountForPercentage) ?? ""
+struct FixedAmountAddCustomAmountInputTypeViewModelAdapter: AddCustomAmountInputTypeViewModelAdapter {
+    weak var viewModel: AddCustomAmountViewModel?
```

Comment:
`Do we want the viewModels in the adapter to be weak and nullable? I'm thinking if accepting a common protocol for fixed and percentage adapters would work well, like:`

Good question @iamgabrielma! The main reason why they are weak and nullable is because we would have a circular reference otherwise. Furthermore, we don't need to hide the `AddCustomAmountViewModel` behind an abstraction, because the adapters work for it, and depend on it, so it's ok that they have all the knowledge about that class.